hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java

```diff
@@ -0,0 +1,284 @@
+                Set<Integer> usedColumns = new HashSet<>();
+                for (RelNode input : union.getInputs()) {
+                    usedColumns.addAll(relToWmKeyMapping.getOrDefault(input, Collections.emptyMap()).keySet());
+                }
+
+                // in that case we cannot use any watermarks.
+                if (usedColumns.size() != 1) {
+                    return;
+                }
+
+                int idx = 0;
+                Map<Integer, Byte> refByteMap = relToWmKeyMapping.get(union.getInput(idx++));
+                for (int i = idx; i < union.getInputs().size(); ++i) {
+                    relToWmKeyMapping.put(union.getInput(i), refByteMap);
+                }
+
+                RelVisitor topDownWmKeysAssigner = new TopDownUnionWatermarkKeyAssignerVisitor(
+                        union,
+                        relToWmKeyMapping
+                );
+
+                topDownWmKeysAssigner.go(union);
+//            } else if (node instanceof StreamToStreamJoinPhysicalRel) {
+//                StreamToStreamJoinPhysicalRel join = (StreamToStreamJoinPhysicalRel) node;
+//                Map<RexInputRef, Byte> leftRefByteMap = refToWmKeyMapping.get(join.getLeft());
+//                if (leftRefByteMap == null) {
+//                    throw QueryException.error("Left input of stream-to-stream JOIN doesn't contain watermarks");
+//                }
+//
+//                Map<RexInputRef, Byte> rightRefByteMap = refToWmKeyMapping.get(join.getRight());
+//                if (rightRefByteMap == null) {
+//                    throw QueryException.error("Right input of stream-to-stream JOIN doesn't contain watermarks");
+//                }
+//
+//                Map<RexInputRef, RexInputRef> jointToLeftInputMapping = join.jointRowToLeftInputMapping();
+//                Map<RexInputRef, RexInputRef> jointToRightInputMapping = join.jointRowToRightInputMapping();
+//
+//                Map<RexInputRef, Byte> jointRefByteMap = new HashMap<>();
+//                for (Map.Entry<RexInputRef, RexInputRef> entry : jointToLeftInputMapping.entrySet()) {
+//                    if (leftRefByteMap.get(entry.getValue()) != null) {
+//                        jointRefByteMap.put(entry.getKey(), leftRefByteMap.get(entry.getValue()));
+//                    }
+//                }
+//
+//                for (Map.Entry<RexInputRef, RexInputRef> entry : jointToRightInputMapping.entrySet()) {
+//                    if (rightRefByteMap.get(entry.getValue()) != null) {
+//                        jointRefByteMap.put(entry.getKey(), rightRefByteMap.get(entry.getValue()));
+//                    }
+//                }
+//
+//                assert leftRefByteMap.size() + rightRefByteMap.size() == jointRefByteMap.size();
+//                refToWmKeyMapping.put(join, jointRefByteMap);
+            } else if (node instanceof Join) {
+                // Hash Join and Nested Loop Join just forward watermarks from left input.
```

Comment:
This is not true for s2s join. It also needs to be fixed in `HazelcastRelMdWatermarkedFields`

Reply:
S2S join rel doesn't exist in this PR, so, existing behavior is preserved. But, I prepared the code for S2S above (commented). I may remove it if you want.